packages/grid/x-data-grid/src/components/base/GridBody.tsx

```diff
@@ -57,14 +57,27 @@ function GridBody(props: GridBodyProps) {
 
   const columnHeadersRef = React.useRef<HTMLDivElement>(null);
   const columnsContainerRef = React.useRef<HTMLDivElement>(null);
-  const windowRef = React.useRef<HTMLDivElement>(null);
   const renderingZoneRef = React.useRef<HTMLDivElement>(null);
 
   apiRef.current.columnHeadersContainerElementRef = columnsContainerRef;
   apiRef.current.columnHeadersElementRef = columnHeadersRef;
-  apiRef.current.windowRef = windowRef; // TODO rename, it's not attached to the window anymore
   apiRef.current.renderingZoneRef = renderingZoneRef; // TODO remove, nobody should have access to internal parts of the virtualization
 
+  const targetRef = React.useCallback(
+    (node: HTMLDivElement) => {
+      if (node && apiRef.current.windowRef?.current == null) {
+        apiRef.current.windowRef = { current: node };
+        const unsubscribe = apiRef.current.subscribeEvent('stateChange', () => {
+          setTimeout(() => {
+            apiRef.current.publishEvent('gridReady');
+          }, 5);
```

Comment:
Regarding the original issue https://github.com/mui/mui-x/issues/6411 specifically, would `gridReady` solve it if the grid is rendered in a loading state while waiting for rows data to load?

Reply:
yes, that's why I add for 'stateChange' listener. if the rows data did not load it won't be fired. the downside is if you create an empty DataGrid, 'gridReady' won't be triggered.

Reply:
here is an example. https://codesandbox.io/s/laughing-noyce-rm95xy?file=/index.tsx